### PR TITLE
top: Replace MICROPY_PORT_BUILTIN_MODULES with MP_REGISTER_MODULE.

### PR DESCRIPTION
**Pull request by jimmo**
No need to have two different ways of registering modules. Simplifies mpconfigport.h and makes it consistent across all ways of defining modules (built-in, extmod, port, board, user-c-module).

It might be good to extend makemodulesdefs.py to also handle module constants (and maybe root pointers?) in the future.

### CODE REVIEW

File: py/makemoduledefs.py [hunk 1]
```diff
@@ -59,18 +59,21 @@ def generate_module_table_header(modules):
...
     print("// Automatically generated by makemoduledefs.py.\n")
     for module_name, obj_module, enabled_define in modules:
         mod_def = "MODULE_DEF_{}".format(module_name.upper())
-        mod_defs.append(mod_def)
+        mod_defs.add(mod_def)
         print(
             (
                 "#if ({enabled_define})\n"
                 "    extern const struct _mp_obj_module_t {obj_module};\n"
+                "    #undef {mod_def}\n"
```
Comment (dlech) [inline py/makemoduledefs.py:71]: Does this mean that if two or more modules with the same name are enabled, then the last one (according to `sorted()`) "wins"?

**Comment by dlech** [comment]
Not a big deal, but is seems like in the vast majority of the cases, the 3rd parameter is 1 or could be 1, so perhaps we should just drop this parameter?

**Comment by dpgeorge** [comment]
> Not a big deal, but is seems like in the vast majority of the cases, the 3rd parameter is 1 or could be 1, so perhaps we should just drop this parameter?

You mean the `#if` config option to enable the module?  That is definitely needed so that ports can selectively enable the modules (if the 3rd param was 1 then all modules would always be enabled on all builds).

**Comment by dlech** [comment]
OK, maybe I misunderstood how the code generator works.

**Comment by jimmo** [comment]
> You mean the #if config option to enable the module? That is definitely needed so that ports can selectively enable the modules (if the 3rd param was 1 then all modules would always be enabled on all builds).

In most cases the `MP_REGISTER_MODULE(..., ..., X)` is already inside an `#if X`, so it could well be just `1`. I followed the existing pattern from modarray.c.

I could try and come up with a varargs macro that makes that last parameter optional (default to 1).

**Comment by dlech** [comment]
> I could try and come up with a varargs macro that makes that last parameter optional (default to 1).

There are probably better ways to spend your time. :wink: 


> In most cases the `MP_REGISTER_MODULE(..., ..., X)` is already inside an `#if X`, so it could well be just `1`. I followed the existing pattern from modarray.c.

This is why I thought we could just eliminate the parameter altogether. But if there are a few special cases where this doesn't work and can't be worked around by adding a `#if` then we need the 3rd parameter.

**Comment by dpgeorge** [comment]
Note that `MP_REGISTER_MODULE()` is not scanned if the file it is in is #include'd.  Eg defining a custom board that has a file `board.c` that has:
```c
#include "common_board_files/modfoo.c`
...
```
And then `modfoo.c` uses `MP_REGISTER_MODULE(foo)`, this will not be picked up.  At the moment the way around this issue is to add another `MP_REGISTER_MODULE(foo)` in `board.c` (in this example).

**Comment by jimmo** [comment]
I've rebased this and updated the renesas-ra port to use it.

> Note that MP_REGISTER_MODULE() is not scanned if the file it is in is #include'd.

\*blinks\*

...I guess that's a good argument for making `makemoduledefs.py` run on the qstr pre-processed code (like error compression does). I think that's a good idea anyway, and it lets us solve the third-argument problem that was discussed above. I will do that in a follow-up PR.

File: ports/nrf/boards/microbit/modules/boardmodules.mk [hunk 2]
```diff
@@ -1,7 +1,6 @@
 BOARD_MICROBIT_DIR = boards/microbit/modules
 
 INC += -I./$(BOARD_MICROBIT_DIR)
-CFLAGS += -DBOARD_SPECIFIC_MODULES
```
Comment (dpgeorge) [inline ports/nrf/boards/microbit/modules/boardmodules.mk:4]: I think this still needs to be defined... the microbit module defines `board_modules_init0()` which still needs to be called from main.

Actually it's a bit confusing what should happen here.  There are two `boardmodules.h` files (one deleted by this PR, another in the pca10059 board dir).

File: ports/stm32/mpconfigport.h [hunk 3]
```diff
@@ -206,35 +206,6 @@ extern const struct _mp_obj_module_t mp_module_onewire;
 #define MACHINE_BUILTIN_MODULE_CONSTANTS
 #endif
 
-#if MICROPY_PY_UTIME
```
Comment (dpgeorge) [inline ports/stm32/mpconfigport.h:209]: there are a bunch of `extern const struct _mp_obj_module_t ...` above that can now be deleted

Comment (jimmo) [reply ports/stm32/mpconfigport.h:209]: Done.

File: ports/nrf/boards/microbit/modules/boardmodules.mk (hunk 2 above)
Comment (jimmo) [reply ports/nrf/boards/microbit/modules/boardmodules.mk]: Ah I missed the dual-purpose here. I have done the simplest change, we should revisit board_init in the future and make it work consistently with the other ports.

**Comment by dpgeorge** [comment]
Thanks for updating, now merged.

**Comment by robert-hh** [comment]
@jimmo As consequence of this and the previous related change, ports which use lwip do not have a usocket module any more registered, causing some tests and uasyncio/stream.py to fail.

**Comment by jimmo** [comment]
> @jimmo As consequence of this and the previous related change, ports which use lwip do not have a usocket module any more registered, causing some tests and uasyncio/stream.py to fail.

@robert-hh Thanks. I will take a look ASAP.

**Pull request merged**